File: September/09 09 25/combinationsum.py
#O(n^2)
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def backtrack(start: int, path: List[int], total: int):
            if total == target:
                res.append(path[:])
                return
            if total > target:
                return

            for i in range(start, len(candidates)):
                path.append(candidates[i])
                # not i + 1 because we can reuse the same element
                backtrack(i, path, total + candidates[i])
                path.pop()

        backtrack(0, [], 0)
        return res

    def combinationSumOneLine(self, candidates, target):
        return [[c] + rest for i, c in enumerate(candidates) for rest in self.combinationSum(candidates[i:], target - c)] if target > 0 else ([[]] if target == 0 else [])

File: September/9/combinationsum.cs
using System;
using System.Collections.Generic;

public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        var res = new List<IList<int>>();
        Backtrack(candidates, target, 0, new List<int>(), res);
        return res;
    }

    private void Backtrack(int[] candidates, int target, int start, List<int> path, IList<IList<int>> res) {
        if (target == 0) {
            res.Add(new List<int>(path));
            return;
        }

        for (int i = start; i < candidates.Length; i++) {
            if (candidates[i] <= target) {
                path.Add(candidates[i]);
                Backtrack(candidates, target - candidates[i], i, path, res); // can reuse same element
                path.RemoveAt(path.Count - 1);
            }
        }
    }
}

public class Program {
    public static void Main() {
        var sol = new Solution();
        int[] candidates = {2, 3, 6, 7};
        var res = sol.CombinationSum(candidates, 7);

        foreach (var combo in res) {
            Console.WriteLine(string.Join(",", combo));
        }
    }
}
